### ISSUE TITLE
socket.create_connection needs a configurable timeout

### ISSUE THREAD

**Issue by dean-82** [eclipse/paho.mqtt.python#426]
I am using this for an embedded project where the broker can disappear and come back due to sleep processing.  When it comes back up I traced back a large reconnection delay to the fact that the call to socket.create_connection is not provided a timeout thereby using the default which can be rather large (observed 20 seconds on Windows 10 with Python 3.7.3).  I want to be able to configure my timeout so it is passed into the call to speed up reconnect attempts.  While I can use socket.setdefaulttimeout(), that is a global value that will affect other socket connections as well.

**Comment by ruoshui1314** [comment]
Maybe #425 is what you want.But waiting for merge.

**Comment by dean-82** [comment]
That is close although I'm not sure that you want to use the keep alive time for that timeout.  Basically, as the keep alive time could be set to a large value, I think that it would be better to have a separate settable timeout value for the initial connection attempt that could be much smaller than the keep alive.  You could initialize it to None so that by default, python will use it's default.

**Comment by ruoshui1314** [comment]
Yeah, I read the source code of paho.mqtt.c about connecting timeout.The default timeout is 10 seconds and the library does not support interface or attribute to set it.
I agree that we may need to separate timeout and keep alive. Whether the library for python should support another parameter for timeout or just give default value, I do not know which is better.
For your issue, maybe support set timeout is better.

**Comment by ruoshui1314** [comment]
I read the source code and find that socket timeout is set to the value of keep alive when connecting to the websocket server or using ssl protocal.What about stay the same? Another suggestion is that update all the socket timeout value to another variable as we discussed before.

**Comment by dean-82** [comment]
That's what I switched to and it is working now.  The main issue is that
this sets the default timeout for all socket connections which is not
necessarily desirable.  If there is a separate timeout in this code, then
it will only be set for the one connection.

**Comment by PierreF** [comment]
I've merged #425 but I won't yet close this issue as from the description the timeout should be configurable (esp. to much lower value from keepalive).

**Comment by aw-snps** [comment]
As a quite happy user of Paho 1.4, I've noticed that #425 became part of the new public 1.5.0 release.

I indeed recommend to refine this implementation in the next release.
We use Paho with setting keepalive==0, which according to the MQTT standard (and properly in the Paho 1.4 implementation) disables the watchdog mechanism.
However, it seems that re-using the Zero as timeout on the socket doesn't allow us to connect at all anymore (errno 115).

The least thing I'd recommend here is to provide the global default timeout again in case keepalive==0. Better, as indicated above already, would be to use separate deadlines for the quite different cases...

Thanks!
